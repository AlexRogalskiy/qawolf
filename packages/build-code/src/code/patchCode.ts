import { buildStepsCode } from "../build";
import { getIndentation, indent, removeLinesIncluding } from "../format";

export const PATCH_SYMBOL = "// 🐺 CREATE CODE HERE";

export const buildPatch = () => {
  let patch = buildStepsCode({
    startIndex,
    steps,
    isTest
  });

  // include the symbol to replace later
  patch += PATCH_SYMBOL;

  return patch;
};

export const canPatch = (code: string) => {
  return code.includes(PATCH_SYMBOL);
};

export const indentPatch = (code: string, patch: string) => {
  /**
   * Match the indentation of the symbol.
   */
  const numSpaces = getIndentation(code, PATCH_SYMBOL);

  return indent(patch, numSpaces, 1);
};

export const patchCode = ({ code, startIndex, steps, isTest }) => {
  let patch = buildPatch();

  patch = indentPatch(code, patch);

  const patchedCode = code.replace(PATCH_SYMBOL, patch);

  if (removeSymbol) {
    code = removeLinesIncluding(code, PATCH_SYMBOL);
  }

  return patchedCode;
};
