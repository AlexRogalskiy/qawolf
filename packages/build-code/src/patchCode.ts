import { getIndentation, indent, removeLinesIncluding } from "./format";

type PatchOptions = {
  code: string;
  patch: string;
};

export const PATCH_HANDLE = "// 🐺 CREATE CODE HERE";

export const canPatch = (code: string) => {
  return code.includes(PATCH_HANDLE);
};

export const indentPatch = (code: string, patch: string) => {
  /**
   * Match the indentation of the symbol.
   */
  const numSpaces = getIndentation(code, PATCH_HANDLE);

  return indent(patch, numSpaces, 1);
};

export const patchCode = ({ code, patch }: PatchOptions) => {
  if (!canPatch(code)) {
    throw new Error("Code missing patch symbol");
  }

  const patchedCode = code.replace(PATCH_HANDLE, indentPatch(code, patch));

  return patchedCode;
};

export const removePatchHandle = (code: string) => {
  return removeLinesIncluding(code, PATCH_HANDLE);
};
