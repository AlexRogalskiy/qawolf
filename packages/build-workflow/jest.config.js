const base = require("../../jest.config.base.js");

module.exports = {
  ...base,
  displayName: "@qawolf/build-workflow",
  name: "@qawolf/build-workflow"
};
