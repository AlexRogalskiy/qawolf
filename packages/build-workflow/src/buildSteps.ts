import { ElementEvent, Step } from "@qawolf/types";
import { concat, sortBy } from "lodash";
import { buildClickSteps } from "./buildClickSteps";
import { buildScrollSteps } from "./buildScrollSteps";
import { buildSelectSteps } from "./buildSelectSteps";
import { buildTypeSteps } from "./buildTypeSteps";

export type BuildStepsOptions = {
  canChange?: boolean;
  events: ElementEvent[];
  stepStartIndex?: number;
};

export const buildSteps = ({
  events,
  ...options
}: BuildStepsOptions): Step[] => {
  const unorderedSteps = concat(
    buildClickSteps(events),
    buildScrollSteps(events),
    buildSelectSteps(events),
    buildTypeSteps(events)
  );

  let steps = sortBy(
    unorderedSteps,
    // ordered by the event index
    step => step.index
  );

  if (!options.canChange) {
    steps = steps.filter(step => !step.canChange);
  }

  // reindex
  const startIndex = options.stepStartIndex || 0;
  steps = steps.map((step, index) => ({
    ...step,
    index: index + startIndex
  }));

  return steps;
};
