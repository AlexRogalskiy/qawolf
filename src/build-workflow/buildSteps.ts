import { concat, sortBy } from 'lodash';
import { buildClickSteps } from './buildClickSteps';
import { buildFillSteps } from './buildFillSteps';
import { buildPressSteps } from './buildPressSteps';
import { buildScrollSteps } from './buildScrollSteps';
import { buildSelectOptionSteps } from './buildSelectOptionSteps';
import { ElementEvent, Step } from '../types';

export const buildSteps = (events: ElementEvent[]): Step[] => {
  const unorderedSteps = concat(
    buildClickSteps(events),
    buildFillSteps(events),
    buildPressSteps(events),
    buildScrollSteps(events),
    buildSelectOptionSteps(events),
  );

  let steps = sortBy(
    unorderedSteps,
    // ordered by the event time
    (step) => step.event.time,
  );

  // reindex
  steps = steps.map((step, index) => ({
    ...step,
    index,
  }));

  return steps;
};
