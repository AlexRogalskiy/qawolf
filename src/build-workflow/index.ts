export * from './buildClickSteps';
export * from './buildKeySteps';
export * from './buildScrollSteps';
export * from './buildSelectOptionSteps';
export * from './buildSteps';
