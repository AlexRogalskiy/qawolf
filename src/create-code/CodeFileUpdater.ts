import { CodeUpdater } from './CodeUpdater';
import { pathExists, readFile, remove, outputFile } from 'fs-extra';

const loadCodeFile = async (path: string): Promise<string> => {
  const codeExists = await pathExists(path);
  if (!codeExists) throw new Error(`No code found at ${path}`);
  return readFile(path, 'utf8');
};

export class CodeFileUpdater extends CodeUpdater {
  private _path: string;
  protected _initialCode: string;

  public static async create(path: string): Promise<CodeFileUpdater> {
    const updater = new CodeFileUpdater(path);
    // TODO some sort of env variable to know if this is a new file and not to set this...
    updater._initialCode = await loadCodeFile(path);
    return updater;
  }

  public constructor(path: string) {
    super();
    this._path = path;
  }

  protected async _loadCode(): Promise<string> {
    return loadCodeFile(this._path);
  }

  protected async _updateCode(code: string): Promise<void> {
    await outputFile(this._path, code, 'utf8');
  }

  public async discard(): Promise<void> {
    this._locked = true;

    if (this._initialCode) {
      await this._updateCode(this._initialCode);
    } else {
      await remove(this._path);
    }
  }

  public path(): string {
    return this._path;
  }
}
