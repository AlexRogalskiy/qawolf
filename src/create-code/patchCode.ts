import { getIndentation, indent, removeLinesIncluding } from './format';

type PatchOptions = {
  code: string;
  patchLines: string[];
  replaceLines: string[];
};

export const CREATE_HANDLE = 'qawolf.create()';

export const PATCH_HANDLE = '// 🐺 create code here';

export const canPatch = (code: string): boolean => {
  return code.includes(PATCH_HANDLE);
};

export const indentPatch = (code: string, patch: string): string => {
  /**
   * Match the indentation of the symbol.
   */
  const numSpaces = getIndentation(code, PATCH_HANDLE);

  return indent(patch, numSpaces, 1);
};

export const patchCode = ({ code, patchLines, replaceLines }: PatchOptions): string => {
  if (!canPatch(code)) {
    throw new Error('Cannot patch without handle');
  }

  const replaceLinesWithHandle = [...(replaceLines || []), PATCH_HANDLE];
  const indentedReplaceLines = indentPatch(code, replaceLinesWithHandle.join('\n'));

  const patchLinesWithHandle = [...(patchLines || []), PATCH_HANDLE];
  const indentedPatch = indentPatch(code, patchLinesWithHandle.join('\n'));

  return code.replace(indentedReplaceLines, indentedPatch);
};

export const removePatchHandle = (code: string): string => {
  return removeLinesIncluding(code, PATCH_HANDLE);
};
