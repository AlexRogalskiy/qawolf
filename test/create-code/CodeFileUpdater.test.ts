import { outputFile, pathExists, readFile } from 'fs-extra';
import { buildSteps } from '../../src/build-workflow/buildSteps';
import { CodeFileUpdater } from '../../src/create-code/CodeFileUpdater';
import { PATCH_HANDLE } from '../../src/create-code/patchCode';

// require manually since fs is mocked
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { events } = require('../fixtures/login.json');
const steps = buildSteps({ events });

jest.mock('fs-extra');

(pathExists as jest.Mock).mockResolvedValue(true);

const mockedReadFile = readFile as jest.Mock;
const mockedOutputFile = outputFile as jest.Mock;

describe('CodeFileUpdater', () => {
  beforeEach(() => {
    mockedOutputFile.mockReset();
  });

  describe('discard', () => {
    it('reverts to initial code', () => {
      // TODO
    });

    it('removes code if there was not pre-existing code', () => {
      // TODO
    });
  });

  describe('finalize', () => {
    it('removes the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `someCode();\n${PATCH_HANDLE}\notherCode();`,
      );
      const updater = new CodeFileUpdater('somepath');
      await updater.finalize();

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe('someCode();\notherCode();');
    });
  });

  describe('prepare', () => {
    it('replaces qawolf.create call with the patch handle', async () => {
      mockedReadFile.mockResolvedValue(
        `  someCode();\n  await qawolf.create();\n  otherCode();`,
      );
      const updater = new CodeFileUpdater('somepath');
      await updater.prepare();

      const updatedFile = mockedOutputFile.mock.calls[0][1];
      expect(updatedFile).toBe(
        `  someCode();\n  // 🐺 CREATE CODE HERE\n  otherCode();`,
      );
    });

    it('throws an error if qawolf.create is not found', async () => {
      mockedReadFile.mockResolvedValue('no create');
      const updater = new CodeFileUpdater('somepath');
      // eslint-disable-next-line jest/valid-expect
      expect(updater.prepare()).rejects.toThrowError();
    });
  });

  describe('update', () => {
    it('saves code with new steps', async () => {
      const updater = new CodeFileUpdater('somepath');

      mockedReadFile.mockResolvedValue(
        `  someCode();\n  ${PATCH_HANDLE}\n  otherCode();`,
      );

      // emulate only the first two steps being ready
      await updater.update({ steps: steps.slice(0, 2) });

      const codeUpdateOne = mockedOutputFile.mock.calls[0][1];
      expect(codeUpdateOne).toMatchSnapshot();
      mockedReadFile.mockResolvedValueOnce(codeUpdateOne);

      await updater.update({ steps });
      const codeUpdateTwo = mockedOutputFile.mock.calls[1][1];
      expect(codeUpdateTwo).toMatchSnapshot();
    });
  });
});
