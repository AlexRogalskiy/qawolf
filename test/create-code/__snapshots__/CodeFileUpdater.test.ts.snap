// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeFileUpdater update saves code with new steps 1`] = `
"  someCode();
  await page.click(selectors[0]);
  await page.click(selectors[1]);
  // üê∫ CREATE CODE HERE
  otherCode();"
`;

exports[`CodeFileUpdater update saves code with new steps 2`] = `
"  someCode();
  await page.click(selectors[0]);
  await page.click(selectors[1]);
  await page.type(selectors[2], \\"tomsmith\\");
  await page.press(selectors[3], \\"Tab\\");
  await page.type(selectors[4], \\"SuperSecretPassword!\\");
  await page.press(selectors[5], \\"Enter\\");
  await page.click(selectors[6]);
  // üê∫ CREATE CODE HERE
  otherCode();"
`;
